flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,343 @@
+  }
+
+  /** Bin value access function
+    *
+    * @param bin bin number to access
+    * @return `v_bin` = value of bin
+    */
+  private[statistics] def getValue(bin: Int): Double = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._1
+  }
+
+  /** Returns the lower limit on values of the histogram
+    *
+    * @return lower limit on values
+    */
+  private def lower: Double = {
+    min
+  }
+
+  /** Returns the upper limit on values of the histogram
+    *
+    * @return upper limit on values
+    */
+  private def upper: Double = {
+    max
+  }
+
+  /** Checks whether the values loaded into data by loadData are properly sorted or not.
+    * All values should be in (lower,upper)
+    * All counters loaded into data by loadData should be positive
+    *
+    */
+  private def checkSanity: Boolean = {
+    val size = bins
+    if (size == 0) return true
+    if (lower >= getValue(0)) return false
+    if (upper <= getValue(size - 1)) return false
+    for (i <- 0 to size - 2) {
+      if (getValue(i + 1) < getValue(i)) return false // equality will get merged later on
+      if (getCounter(i) <= 0) return false
+    }
+    if (getCounter(size - 1) <= 0) return false
+    // now bring the histogram to capacity and merge all <i>very-close</i> elements
+    while (mergeElements()) {}
+    true
+  }
+
+  /** Searches for the index with value just greater than p
+    * If `p >= v_{bins-1}`, return bins
+    *
+    * @param p value to search for
+    * @return the bin with value just greater than p
+    */
+  private def find(p: Double): Int = {
+    val size: Int = bins
+    if (size == 0) {
+      return 0
+    }
+    if (p < getValue(0)) {
+      return 0
+    }
+    for (i <- 0 to size - 2) {
+      if (p >= getValue(i) && p < getValue(i + 1)) {
+        return i + 1
+      }
+    }
+    size
+  }
+
+  /** Merges the closest two elements in the histogram
+    * If we're over capacity, merge without thought
+    * Otherwise, only merge when two elements have almost equal values. We're talking doubles, so
+    * no exact equalities
+    *
+    * @return Whether we succeeded in merging any two elements
+    */
+  private def mergeElements(): Boolean = {
+    val size: Int = bins
+    var minDiffIndex: Int = -1
+    var minDiff: Double = MaxValue
+    for (i <- 0 to size - 2) {
+      val currDiff: Double = getValue(i + 1) - getValue(i)
+      if (currDiff < minDiff) {
+        minDiff = currDiff
+        minDiffIndex = i
+      }
+    }
+    if (bins > capacity || minDiff < 1e-9) {
+      val weightedValue = getValue(minDiffIndex + 1) * getCounter(minDiffIndex + 1) +
+        getValue(minDiffIndex) * getCounter(minDiffIndex)
+      val counterSum = getCounter(minDiffIndex + 1) + getCounter(minDiffIndex)
+      data.update(minDiffIndex, (weightedValue / counterSum, counterSum))
+      data.remove(minDiffIndex + 1)
+      return true
+    }
```

Comment:
Saving the return value of `getValue` method and `getCounter` method into immutable variables and reusing them would be better.